### ISSUE TITLE
🖼️ Responsive image generated sizes

### ISSUE THREAD

**Issue by dbismut** [withastro/astro#12823]
### Astro Info

```block
Astro                    v5.1.1
Node                     v18.20.3
System                   Linux (x64)
Package Manager          unknown
Output                   static
Adapter                  none
Integrations             none
```

### Describe the Bug

I've played with the `responsiveImages` experimental flag set to `true` and `layout="responsive"` a few minutes and it seems that _all_ that this can do is taking the natural size of the image and generate sizes such as `sizes="(min-width: 2302px) 2302px, 100vw"` no matter how  big the image is actually rendered. Am I missing something?

### What's the expected result?

I naively thought that an image that was `50%` of the width in desktop and 100% of the width in mobile would have the sizes magically set to `"100vw, (min-width: 640px) 50vw"`.

I was really curious as to how this could actually work and what was the magic behind it (I thought you would use some sort of clever headless rendering somehow eventhough perf would have been dramatic I guess).

Please disregard if this is the intended behavior, but it's not how sizes should work in order to reduce payload and improve core vitals and I believe you should mention it in the docs.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-neb689p7

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by ascorbic** [comment]
No, because as you state it has no way of knowing when it builds what width it will be if you're constraining it with CSS. Your best bet there is to set the image `width` to whatever the maximum is that you want it to be displayed (e.g. 50% of your max column width). That will mean it will use that as the maximum displayed width rather than the intrinsic size and will generate the sizes and srcset accordingly. You can optionally set `sizes` manually too if it will be constrained to 50% too, but that won't affected the generated image sizes.

**Comment by dbismut** [comment]
Ok so the feature only apply to full width images. Don't you think this is worth mentioning in the docs?

 As as a side note, srcset being always the same is fine, what's really hard in userland is nailing the right sizes attribute.

**Comment by ascorbic** [comment]
No, it's not just full width images, but it does assume that when the window is resized enough to limit the size of the image then it will be full width at that point, because most responsive designs work like that. You shouldn't be relying on the responsive resizing to fit it in your container on a larger page: you should set the width.

**Comment by dbismut** [comment]
Thanks for keeping the discussion open. I don't want to be picky but a fixed width isn't the correct way to handle responsive layouts either; when an image is responsive, its width is more often than not a fraction of the viewport. With the generated sizes given by the experimental `responsiveImage`, the image downloaded src will always default to the full width of the viewport (100vw) whenever the viewport is lower than the given width or the image intrinsic width. So you're are de facto kinda assuming that an image is full width down a certain point.
